tools/cleanup/cleanup.go

```diff
@@ -0,0 +1,308 @@
+			} else {
+				region, zone := gke.RegionZoneFromLoc(cluster.Location)
+				if err := gkeClient.DeleteCluster(project, region, zone, cluster.Name); err != nil {
+					return count, errors.Wrapf(err, "error deleting cluster %q in project %q", cluster.Name, project)
+				}
+				count++
+			}
+		}
+	}
+	return count, nil
+}
+
+// deleteImages deletes old docker images from a given project.
+func deleteImages(project string, before time.Time) (int, error) {
+	repoRoot := *gcr + "/" + project
+	// TODO(adrcunha): This should be a helper function, like https://github.com/knative/pkg/blob/6d806b998379948bd0107d77bcd831e2bdb4f3cb/testutils/clustermanager/e2e-tests/gke.go#L281
+	if repoRoot == "gcr.io/knative-releases" || repoRoot == "gcr.io/knative-nightly" {
+		return 0, fmt.Errorf("cleaning up %q is forbidden", repoRoot)
+	}
+	gcrrepo, err := name.NewRepository(repoRoot)
+	if err != nil {
+		return 0, errors.Wrapf(err, "cannot open registry %q", repoRoot)
+	}
+	count := 0
+	// Walk down the registry, checking all images and deleting the old ones.
+	return count, google.Walk(gcrrepo, func(repo name.Repository, tags *google.Tags, err error) error {
+		// If we got an error, just return it, there's nothing to do here.
+		if tags == nil {
+			return err
+		}
+		for k, m := range tags.Manifests {
+			ref := repo.String() + "@" + k
+			age := int(time.Since(m.Uploaded).Hours() / 24)
+			log.Printf("%q is %d days old", ref, age)
+			if m.Uploaded.Before(before) {
+				if *dryRun {
+					log.Printf("[DRY RUN] delete %q", ref)
+				} else {
+					// Delete all tags first, otherwise the image can't be deleted.
+					for _, tag := range m.Tags {
+						if err := deleteImage(repo.String() + ":" + tag); err != nil {
+							return err
+						}
+					}
+					if err := deleteImage(ref); err != nil {
+						return err
+					}
+					log.Printf("Deleted %q (uploaded on %s)", ref, m.Uploaded)
+					count++
+				}
+			}
+		}
+		return nil
+	}, google.WithAuthFromKeychain(auther))
+}
+
+// deleteResources call resourceDeleter functions in parallel, one for each given project.
+func deleteResources(projects []string, before time.Time, deleteFunc resourceDeleter) (int, []string) {
+	// Map of errors returned from functions, to avoid repeated entries.
+	var errs = make(map[string]error)
+
+	// Locks for the concurrent tasks.
+	var errsLock sync.RWMutex
+	var countLock sync.RWMutex
+	var wg sync.WaitGroup
+
+	// Blocking channel to keep concurrency under control.
+	concurrencyChan := make(chan struct{}, *concurrentOperations)
+	defer close(concurrencyChan)
+
+	count := 0
+	for i := range projects {
+		wg.Add(1)
+		go func(project string) {
+			defer wg.Done()
+			// Block processing if concurrency reached the limit.
+			concurrencyChan <- struct{}{}
+			// Do not process if previous invocations failed. This prevents a large
+			// build-up of failed requests and rate limit exceeding (e.g. bad auth).
+			errsLock.RLock()
+			failed := len(errs) > 0
```

Comment:
so a single failure fails all following events?

Reply:
Yes. As explained in the comment above, we want to avoid an avalanche of errors (usually the quota error due to the amount of projects).